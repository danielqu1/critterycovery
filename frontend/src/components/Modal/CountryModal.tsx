// Modal acts as an instance page for countries. Displays information and media for countries
import {useEffect, useState} from 'react'; 
import {Modal, Button, Container, Col, Row, ListGroup} from 'react-bootstrap' 
import axios from 'axios'
import { Image } from 'antd';

import loadGIF from '../../images/loading.gif'
import defaultImage from '../../data/DefaultImage'

interface species{
	scientific_name: string;
}

interface habitats{
	name: string;
}

const no_info = "information not available"

function CountryModal(props: any) {
	const [species, setSpecies] = useState(new Array<species>());
	const [habitats, setHabitats] = useState(new Array<habitats>());

	// Loads the connection data between country and species
	useEffect(() => {
	setSpecies(new Array<species>())
	if(props.country != null){
		axios.get('/api/countries/species/name='+props.country.name).then((response) => {
		setSpecies(response.data.species);
		})
	}
	}, [props.country]);

	// Loads the connection data between country and habitat
	useEffect(() => {
	setHabitats(new Array<habitats>())
	if(props.country != null){
		axios.get('/api/countries/habitats/name='+props.country.name).then((response) => {
			setHabitats(response.data.habitats);
		})
	}
	}, [props.country]);

	if(props.country == null){
	return(<></>)
	}

	// Build lists of connections with links
	const speciesLinks = [];
	for (let i = 0; i < species.length; i++) {
	speciesLinks.push(<a style={{ cursor: 'pointer' }} href={'/species/'+species[i].scientific_name}>{species[i].scientific_name+' '}</a>);
	}
	if (species.length === 0) {
		speciesLinks.push(<>{no_info}</>)
	}
	speciesLinks.push(<br/>)

	const habitatLinks = [];
	for (let i = 0; i < habitats.length; i++) {
	habitatLinks.push(<a style={{ cursor: 'pointer' }} href={'/habitats/'+habitats[i].name}>{habitats[i].name+' '}</a>);
	}
	if (habitats.length === 0) {
		habitatLinks.push(<>{no_info}</>)
	}
	habitatLinks.push(<br/>)
	
	return (
	<Modal
		{...props}
		size="lg"
		aria-labelledby="contained-modal-title-vcenter"
		centered
	>
		<Modal.Header closeButton>
		<Modal.Title id="contained-modal-title-vcenter">
			{props.country.name}
		</Modal.Title>
		</Modal.Header>
		<Modal.Body>
		<Container style={{marginBottom: '4%'}}>
				<Row xs={1} lg={2}>
					<Col style={{paddingRight:'2%'}}>
						<Row style={{marginBottom:'4%'}}>
							<Image 
								width='100%'
								height='100%'
								style={{objectFit:'cover'}}
								alt={"Flag of "+props.country.name}
								src={props.country.flag}
								placeholder={
									<Image
									preview={false}
									src={loadGIF}
									width='100%'
									height='100%'
									style={{objectFit:'cover'}}
									alt="loading gif"
									/>
								}
								preview={false}
								fallback={defaultImage()}
							/><br/>
						</Row>
						<Row>
							<ListGroup>
								<ListGroup.Item style={{backgroundColor:'#ff8368'}}><span style={{ float: "left" }}>Capital: {props.country.capital ? props.country.capital : "none"}</span></ListGroup.Item>
								<ListGroup.Item style={{backgroundColor:'#FFB37C'}}><span style={{ float: "left" }}>Total Population: {props.country.total_pop ? props.country.total_pop : no_info}</span></ListGroup.Item>
								<ListGroup.Item style={{backgroundColor:'#fff868'}}><span style={{ float: "left" }}>Area: {props.country.area ? props.country.area : no_info} km&sup2;</span></ListGroup.Item>
								<ListGroup.Item style={{backgroundColor:'#689dff'}}><span style={{ float: "left" }}>Gini Index: {props.country.gini_index ? props.country.gini_index : "none"}</span></ListGroup.Item>
								<ListGroup.Item style={{backgroundColor:'#68ffa2'}}><span style={{ float: "left" }}>Region: {props.country.region ? props.country.region : "none"}</span></ListGroup.Item>
								<ListGroup.Item style={{backgroundColor:'#fff868'}}><span style={{ float: "left" }}>Subregion: {props.country.subregion ? props.country.subregion : "none"}</span></ListGroup.Item>
								
							</ListGroup>
						</Row>
					</Col>
					<Col style={{paddingLeft:'6%'}}>
						<Row>
							<ListGroup>
								<ListGroup.Item style={{backgroundColor: '#cdedff'}}><span style={{ float: "left" }}>Species: {speciesLinks}</span></ListGroup.Item>
								<ListGroup.Item style={{backgroundColor: '#FFF8B4'}}><span style={{ float: "left" }}>Habitats: {habitatLinks}</span></ListGroup.Item>
							</ListGroup>
						</Row>
						<Row style={{marginTop:'2%'}}>
							<iframe title="Map of Country"
								width="100%"
								height="400"
								style={{border: 0}}
								loading="lazy"
								allowFullScreen
								src={props.country.embedded_map_link}>
							</iframe> <br/>
							Latitude: {props.country.latitude ? props.country.latitude : no_info }&deg; 
							Longitude: {props.country.longitude ? props.country.longitude : no_info}&deg;<br/>
							Alpha2: {props.country.alpha2_code} Alpha3: {props.country.alpha3_code}<br/>
						</Row>
					</Col>
				</Row>
			</Container>

		</Modal.Body>
		<Modal.Footer>
		<Button onClick={props.onHide}>Close</Button>
		</Modal.Footer>
	</Modal>
	);
}

export default CountryModal;