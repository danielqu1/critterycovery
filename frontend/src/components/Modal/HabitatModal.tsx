import React from 'react'; 
import {Modal, Button, Image} from 'react-bootstrap'
import axios from 'axios'

interface species{
  scientific_name: string;
}

interface country{
  name: string;
  alpha3_code: string;
}

function HabitatModal(props: any) {
  const country_default : country = {name: 'United States of America', alpha3_code: 'USA'}
  const [species, setSpecies] = React.useState(new Array<species>());
  const [country, setCountry] = React.useState(country_default);

  React.useEffect(() => {
    if(props.habitat != null){
      axios.get('/api/countries/alpha3_code='+props.habitat.countries).then((response) => {
        setCountry(response.data.country);
      })
    }
  }, [props.habitat]);

  React.useEffect(() => {
    setSpecies(new Array<species>())
    if(props.habitat != null){
      axios.get('/api/countries/species/name='+country.name).then((response) => {
        setSpecies(response.data.species);
      })
    }
  }, [country]);

  if(props.habitat == null){
    return(<a></a>)
  }
  const speciesLinks = [];
  for (let i = 0; i < species.length; i++) {
    speciesLinks.push(<a style={{ cursor: 'pointer' }} href={'/species/'+species[i].scientific_name}>{species[i].scientific_name+' '}</a>);
  }
  speciesLinks.push(<br/>)

  let countryLink = <a></a>;
  if(country != null){
    countryLink = (<a style={{ cursor: 'pointer' }} href={'/countries/'+country.name}>{country.name+' '}</a>)
  }

  return (
    <Modal
      {...props}
      size="lg"
      aria-labelledby="contained-modal-title-vcenter"
      centered
    >
      <Modal.Header closeButton>
        <Modal.Title id="contained-modal-title-vcenter">
          {props.habitat.name}
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Image src={props.habitat.image_link} rounded fluid style={{width:'50%'}}/><br/>
        ID: {props.habitat.id.toString()}<br/>
        Marine: {props.habitat.marine.toString()}<br/>
        Water Area: {props.habitat.reported_marine_area.toString()}<br/>
        Land Area: {props.habitat.reported_terrestrial_area.toString()}<br/>
        icun Category: {(props.habitat.icun_category)? props.habitat.icun_category.toString() : ""}<br/>
        Designation: {props.habitat.designation_name}<br/>
        Link: <a href={props.habitat.link}>{props.habitat.link}</a><br/>
        <iframe title="Map of Country"
          width="100%"
          height="20%"
          style={{border: 0}}
          loading="lazy"
          allowFullScreen
          src={props.habitat.embedded_map_link}>
        </iframe> <br/>
        Country: <br/>{countryLink}<br/>
        Species: <br/>{speciesLinks}<br/>
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={props.onHide}>Close</Button>
      </Modal.Footer>
    </Modal>
  );
}

export default HabitatModal;