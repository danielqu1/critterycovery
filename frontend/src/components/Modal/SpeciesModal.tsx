import {useState, useEffect} from 'react'; 
import {Modal, Button, Image, Card, ListGroup } from 'react-bootstrap'
import axios from 'axios'

interface countries {
	country: string;
	alpha3_code: string;
}

interface habitats {
	name: string;
}

const NA = "N/A";

const no_info = "information not available"

function SpeciesModal(props: any) {

	const [countries, setCountries] = useState(new Array<countries>());
    const [habitats, setHabitats] = useState(new Array<habitats>());

	useEffect(() => {
		setCountries(new Array<countries>())
		if(props.species != null) {
			axios.get('/api/species/countries/name='+props.species.scientific_name).then((response) => {
				setCountries(response.data.countries);
			})
		}
	}, [props.species]);

	useEffect(() => {
		setHabitats(new Array<habitats>())
		if(countries !== null) {
			for (let i = 0; i < countries.length; i++) {
				axios.get('/api/countries/habitats/name='+countries[i].country).then((response) => {
					setHabitats(habitats.concat(response.data.habitats));
				})
			}
		}
    }, [countries, habitats]);
  
	if(props.species == null){
		return(<></>)
    }

	const countryLinks = [];
	for (let i = 0; i < countries.length; i++) {
		countryLinks.push(<a style={{ cursor: 'pointer' }} href={'/countries/'+countries[i].country}>{countries[i].country+' '}</a>);
	}
	if (countries.length === 0) {
		countryLinks.push(<>{no_info}</>)
	}
	countryLinks.push(<br/>)

	const habitatLinks = [];
	for (let i = 0; i < habitats.length; i++) {
		habitatLinks.push(<a style={{ cursor: 'pointer' }} href={'/habitats/'+habitats[i].name}>{habitats[i].name+' '}</a>);
	}
	if (habitats.length === 0) {
		habitatLinks.push(<>{no_info}</>)
	}
	habitatLinks.push(<br/>)
  

	return (
		<Modal
			{...props}
			size="lg"
			aria-labelledby="contained-modal-title-vcenter"
			centered
		>
			<Modal.Header closeButton>
				<Modal.Title id="contained-modal-title-vcenter">
					{props.species.common_name !== NA ? props.species.common_name : props.species.scientific_name}
				</Modal.Title>
			</Modal.Header>
			<Modal.Body>
				<Image src={props.species.image_link} rounded fluid style={{width:'50%'}}/><br/>
				Scientific Name: {props.species.scientific_name}<br />

				<Card style={{ width: '60%' }} bg="light green">
					<Card.Body>
						<Card.Title>Taxonomy</Card.Title>
						<ListGroup variant="flush">
							<ListGroup.Item><span style={{ float: "left" }}>Kingdom: {props.species.kingdom}</span></ListGroup.Item>
							<ListGroup.Item><span style={{ float: "left" }}>Phylum: {props.species.phylum}</span></ListGroup.Item>
							<ListGroup.Item><span style={{ float: "left" }}>Class: {props.species._class}</span></ListGroup.Item>
							<ListGroup.Item><span style={{ float: "left" }}>Order: {props.species._order}</span></ListGroup.Item>
							<ListGroup.Item><span style={{ float: "left" }}>Family: {props.species.family}</span></ListGroup.Item>
							<ListGroup.Item><span style={{ float: "left" }}>Genus: {props.species.genus}</span></ListGroup.Item>
							<ListGroup.Item>Taxonomic Notes: <div dangerouslySetInnerHTML={{ __html: props.species.taxonomic_notes }}></div></ListGroup.Item>
						</ListGroup>
					</Card.Body>
				</Card>
				

				Countries: <br /> {countryLinks} <br/>
				Habitats: <br /> {habitatLinks} <br/>
				Subspecies: {props.species.subspecies ? props.species.subspecies : "none"}<br/>
				Subpopulations: {props.species.subpopulations ? props.species.subpopulations : "none"}<br/>
				Population Trend: {props.species.population_trend ? props.species.population_trend : no_info}<br/>
				Marine: {props.species.marine.toString()}<br/>
				Freshwater: {props.species.freshwater.toString()}<br/>
				Terrestrial: {props.species.terrestrial.toString()}<br/>

				Rationale: <br/><div dangerouslySetInnerHTML={{ __html: props.species.rationale ? props.species.rationale : no_info }}></div><br/>
				Geographic Range: <br/><div dangerouslySetInnerHTML={{ __html: props.species.geographic_range ? props.species.geographic_range : no_info }}></div><br/>
				Population: <br/><div dangerouslySetInnerHTML={{ __html: props.species.population ? props.species.population : no_info }}></div><br/>
				Habitat Description: <br/><div dangerouslySetInnerHTML={{ __html: props.species.habitat ? props.species.habitat : no_info}}></div><br/>
				Threats: <br/><div dangerouslySetInnerHTML={{ __html: props.species.threats ? props.species.threats : no_info }}></div><br/>
				Conservation Measures: <br/><div dangerouslySetInnerHTML={{ __html: props.species.conservation_measures ? props.species.conservation_measures : no_info }}></div><br/>

			</Modal.Body>
			<Modal.Footer>
				<Button onClick={props.onHide}>Close</Button>
			</Modal.Footer>
		</Modal>
	);
}

export default SpeciesModal;