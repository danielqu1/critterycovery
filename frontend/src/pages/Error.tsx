import React from 'react';
import { Table, Nav } from 'react-bootstrap';

function Error() {
    return(
        <div>
            <h1>Error 404. Page not Found.</h1>
        </div>
    );
}

export default Error;
