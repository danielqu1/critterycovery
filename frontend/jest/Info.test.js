import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import "@testing-library/jest-dom/extend-expect"

//import About from "../src/pages/about/About.tsx";     // Jest complains about interface in About.tsx
import Error from "../src/pages/Error.tsx";
import CountryModal from "../src/components/Modal/CountryModal.tsx";
import HabitatModal from "../src/components/Modal/HabitatModal.tsx";
import SpeciesModal from "../src/components/Modal/SpeciesModal.tsx";



configure({ adapter: new Adapter() });

describe("Render views", () => {
    test("Error page", () => {
        const aboutTest = shallow(<Error />)
        expect(aboutTest).toMatchSnapshot()
    });

    test("Country Modal page", () => {
        const aboutTest = shallow(<CountryModal/>)
        expect(aboutTest).toMatchSnapshot()
    });

    test("Habitat Modal page", () => {
        const aboutTest = shallow(<HabitatModal />)
        expect(aboutTest).toMatchSnapshot()
    });

    test("Species Modal page", () => {
        const aboutTest = shallow(<SpeciesModal />)
        expect(aboutTest).toMatchSnapshot()
    });
})
