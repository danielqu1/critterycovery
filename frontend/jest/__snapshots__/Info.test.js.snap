// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render views Country Modal page 1`] = `ShallowWrapper {}`;

exports[`Render views Error page 1`] = `ShallowWrapper {}`;

exports[`Render views Habitat Modal page 1`] = `ShallowWrapper {}`;

exports[`Render views Species Modal page 1`] = `ShallowWrapper {}`;
